File: Inorder Successor in BST[30-July-2023]/Optimal Approach/Solution.java
//  >> JAVA CODE <<

class Solution
{
    // returns the inorder successor of the Node x in BST (rooted at 'root')
    
	public Node inorderSuccessor(Node root,Node x)
    {
       Node ans=null;
       
       while(root!=null)
       {
           if(root.data<=x.data)
           {
               root=root.right;
           }
           else
           {
               ans=root;
               root=root.left;
           }
       }
       
        return ans;
       
    }

}

File: Inorder Successor in BST[30-July-2023]/Optimal Approach/solution.cpp
//  >> CPP CODE <<

class Solution{
  public:
    // returns the inorder successor of the Node x in BST (rooted at 'root')
    Node * inOrderSuccessor(Node *root, Node *x)
    {
        Node *ans=NULL;
        
        while(root!=NULL)
        {
            if(root->data<=x->data)
            {
                root=root->right;
            }
            else
            {
                ans=root;
                root=root->left;
            }
        }
        
            return ans;
    }
};